x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java

```diff
@@ -0,0 +1,275 @@
+        if (size == 0) {
+            throw new IllegalArgumentException("empty set should not be used");
+        }
+
+        Set<String> actions = new HashSet<>();
+        Set<String> patterns = new HashSet<>();
+        for (String name : names) {
+            name = name.toLowerCase(Locale.ROOT);
+            if (isValidPrivilegeName(name) != false) {
+                ApplicationPrivilege privilege = lookup.get(name);
+                if (privilege != null && size == 1) {
+                    return privilege;
+                } else if (privilege != null) {
+                    patterns.addAll(Arrays.asList(privilege.patterns));
+                } else {
+                    throw new IllegalArgumentException("unknown application privilege [" + names + "]");
+                }
+            } else {
+                actions.add(name);
+            }
+        }
+
+        if (actions.isEmpty()) {
+            return new ApplicationPrivilege(application, names, patterns.toArray(new String[patterns.size()]), emptyMap(), true);
+        } else {
+            patterns.addAll(actions);
+            return new ApplicationPrivilege(application, names, patterns.toArray(new String[patterns.size()]), emptyMap(), false);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return application + ":" + super.toString() + "(" + Strings.arrayToCommaDelimitedString(patterns) + ")";
+    }
+
+    @Override
+    public int hashCode() {
+        int result = super.hashCode();
+        result = 31 * result + Objects.hashCode(application);
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return super.equals(o) && Objects.equals(this.application, ((ApplicationPrivilege) o).application);
+    }
+
+    public XContentBuilder toIndexContent(XContentBuilder builder, Params params) throws IOException {
+        return writeXContent(builder, true);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return writeXContent(builder, false);
+    }
+
+    private XContentBuilder writeXContent(XContentBuilder builder, boolean includeType) throws IOException {
+        assert name.size() == 1;
+
+        builder.startObject()
+            .field(Fields.APPLICATION.getPreferredName(), application)
+            .field(Fields.NAME.getPreferredName(), getPrivilegeName())
+            .array(Fields.ACTIONS.getPreferredName(), this.patterns)
+            .field(Fields.METADATA.getPreferredName(), this.metadata);
+
+        if (includeType) {
+            builder.field(Fields.TYPE.getPreferredName(), DOC_TYPE_VALUE);
+        }
+        return builder.endObject();
+    }
+
+    public static ApplicationPrivilege parse(XContentParser parser, boolean allowType) throws IOException {
+        return PARSER.parse(parser, allowType);
+    }
+
+    public static ApplicationPrivilege readFrom(StreamInput in) throws IOException {
```

Comment:
> Can we have a constructor that takes a StreamInput instead of the static method?

Not easily (or at least, not with code that ends up being readable).
The constructors need to call `super` with parameters that are read from the stream. 
There isn't a nice pattern for doing that in a constructor.